[Dynamic Dashboard] Fix issue of not updating the orders count in some scenarios

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

### Description
In some cases, and given the fact that order status options are fetched only on app launch and when switching sites, it could happen that the app shows the performance card for stores with no orders or vice versa.

This PR fixes this by making two changes:
1. It updates the logic of `ObserveProcessingOrdersCount` to always use the order status API for infering the number of processing orders, instead of using the `/wc-analytics/reports/totals` endpoint, this update will allow us to refresh the cache whenever any changes happen.
2. It updates the logic of `ObserveSiteOrdersState` to observe the changes of order status options instead of retrieving a single snapshot.

With these two changes, the dashboard will always observe the updated state.

### Testing instructions
1. Use a store with a single order.
2. Open the app.
3. Notice the performance card is shown.
4. Switch to the order list screen.
5. Delete the order.
6. Go back to the dashboard screen.
7. Notice the performance card was disabled (there could be a small delay for this, and it's because of the delay of the `undo` action that we use before trashing the order).
8. Repeat the test with "adding a new order"
9. Notice the performance card was enabled.

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/1657201/75de6a5a-537d-4fd6-878c-22bac62c5e47



- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
A minor improvement to the logic

hichamboushaba committed:
Always use order status options to infer the number of processing orders

This is done by updating the logic to fetch the order status options instead of the `/wc-analytics/reports/orders/totals`

hichamboushaba committed:
Update logic of ObserveSiteOrderState to observe status options changes

hichamboushaba committed:
Remove redundant fetch

We already do a fetch when the app starts in the SiteObserver, so here we should just use the cached value initially, and then observe changes

hichamboushaba committed:
Improve FakeDispatcher to allow executing a callback when dispatching

hichamboushaba committed:
Fix unit tests

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/ObserveProcessingOrdersCount.kt:
```diff
@@ -42,18 +55,11 @@ class ObserveProcessingOrdersCount @Inject constructor(
         // Start with the cached value
         emit(getCachedValue(site))
 
-        // emit updated value
-        fetchOrdersCount(site)?.let { emit(it) }
```
Since we fetch the order status options on app launch in `SiteObserver`, making another fetch here is redundant, so I removed it.

Now we will start with the cached value, then observe changes and fetch when needed.

0nko reviewed (commented):
I started out with a store with no orders (there are some trashed orders) and the cards are still being shown, even after force-refreshing the orders and the dashboard. 🤔 

[Screen_recording_20240517_095857.webm](https://github.com/woocommerce/woocommerce-android/assets/1522856/fe1627a8-461c-48db-9311-8a587b3f5ed5)

hichamboushaba commented:
@0nko thanks for the test, can you invite me to your store to check this? This doesn't happen on my store, `trash` is not returned as a status option for me, which means the trashed orders are not counted.

hichamboushaba committed:
A minor improvement to avoid redundant API call

We skip the initial orders count, with this we don't make a redundant API call to fetch the order status options, something that SiteObserver already handles.

hichamboushaba committed:
Remove usage of `observeOrdersCount` for detecting if store has orders

The function checks only the local DB, and given how pagination works, we don't remove trashed orders from the API, so when a site has trashed orders, we keep thinking that they have valid orders, and we keep showing the stats cards.

0nko reviewed (approved):
Works well after the last change. Nice fix! :shipit:

hichamboushaba committed:
Merge branch 'trunk' into dashboard/improve-logic-of-has-orders
